### PR TITLE
Fixed "get_arc_center()" method of Arc class

### LINKED ISSUE #480: Sector.get_arc_center() is broken.
Sector.get_arc_center() does not work and always returns a warning message.
```sh
UserWarning: Can't find Arc center, using ORIGIN instead
 warnings.warn("Can't find Arc center, using ORIGIN instead")
```
Above error can be reproduced using the below code.
```py
from manim import *
class Test(Scene):
    def construct(self):
        sec = Sector(start_angle=0, angle=5*PI/3).move_to([1,0,0])
        print(sec.get_arc_center())
```

### PR DESCRIPTION
**Pull request by AkashKarnatak**
## Explaination
`get_arc_center()` method fails if the arc radius is zero. Therefore an if statement has been added to indirectly check whether the arc radius is zero or not.
If the anchor points lie at the same point, it implies that arc radius is zero. Thus arc_center also lie at that point.

This PR fixes #480 

## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

### CODE REVIEW

**Review by leotrs — CHANGES_REQUESTED** [review]
I cannot tell from your explanation, or from the diff, how this is going to solve #480. Can you please be more specific? Also, I would love to add a test for this. Something simple will do, no need to add video tests
```
def some_test():
    assert some_arc.get_arc_center() == ORIGIN
```
Or something.

**Comment by AkashKarnatak** [comment]
`get_arc_center()` fails whenever the radius of arc is zero. I guess you're right, I should change `get_arc_center()` method. Using an if statement I can determine whether the radius of arc is zero or not. If it's zero, anchor points will lie at the same point, otherwise they won't

**Comment by AkashKarnatak** [comment]
@leotrs I have updated my PR, let me know if you want me to change anything else

**Comment by AkashKarnatak** [comment]
> I cannot tell from your explanation, or from the diff, how this is going to solve #480. Can you please be more specific? Also, I would love to add a test for this. Something simple will do, no need to add video tests
> 
> ```
> def some_test():
>     assert some_arc.get_arc_center() == ORIGIN
> ```
> 
> Or something like that.
> 
> EDIT: let me be more specific. I understand what you mean about the inner arc points not being helpful. But why does this change the behavior of get_arc_center? Wouldn't it be better to change get_arc_center to work in the case where there are no inner points?

Which file should I modify to add test case?

**Comment by leotrs** [comment]
Thanks for the update. Can you please create a new file `tests/test_geometry.py`?

**Comment by AkashKarnatak** [comment]
> Thanks for the update. Can you please create a new file `tests/test_geometry.py`?

Done, I have added a simple test case for get_arc_center()

**Comment by behackl** [comment]
Practically, this fix works and it seems to do what it is intended for. However, I think that it would be better to fix this with a more robust and readable solution:

1. Write a custom `__init__` for `AnnularSector` that stores `outer_arc` and `inner_arc` as properties.
2. Introduce `AnnularSector.get_arc_center` as something like
  ```py
  def get_arc_center(self):
      return self.outer_arc.get_arc_center()
  ```
3. Change the check in `Arc.get_arc_center` to a `self.radius == 0`. (Personally, I don't like comparing the first two anchor points that much when an `Arc` knows its radius anyways.)

I would really like to see a cleanup like that, if you are interested in having a look at this strategy. If not, we can stick with the current approach and will delegate cleaning up the codebase to a follow-up PR.

**Comment by AkashKarnatak** [comment]
@behackl I agree, a robust and readable solution would be better. I'll make those changes listed in points 1 and 2. 

I didn't use `self.radius` in the check because it itself is broken. `self.radius` does not update when we scale Arc objects. Although this can be fixed if I override `scale()` in Arc class.

**Comment by AkashKarnatak** [comment]
> Practically, this fix works and it seems to do what it is intended for. However, I think that it would be better to fix this with a more robust and readable solution:
> 
>     1. Write a custom `__init__` for `AnnularSector` that stores `outer_arc` and `inner_arc` as attributes.
> 
>     2. Introduce `AnnularSector.get_arc_center` as something like
> 
> 
> ```python
> def get_arc_center(self):
>     return self.outer_arc.get_arc_center()
> ```
> 
>     1. Change the check in `Arc.get_arc_center` to a `self.radius == 0`. (Personally, I don't like comparing the first two anchor points that much when an `Arc` knows its radius anyways.)
> 
> 
> I would really like to see a cleanup like that, if you are interested in having a look at this strategy. If not, we can stick with the current approach and will delegate cleaning up the codebase to a follow-up PR.

If I store `outer_arc` and `inner_arc` as attributes, I need to make sure that they transform and translate as the original `Mobject` does. 
One way is to store them as `submobjects` using `self.add`. But this way they'll be always be displayed in a scene along with the parent `Mobject`. 
Is there any other solution that you would suggest?

**Comment by behackl** [comment]
> If I store `outer_arc` and `inner_arc` as attributes, I need to make sure that they transform and translate as the original `Mobject` does. 
> One way is to store them as `submobjects` using `self.add`. But this way they'll be always be displayed in a scene along with the parent `Mobject`. 
> Is there any other solution that you would suggest?

Thank you for your efforts! You are right, I did not think of that. In this case, the only possible option of a nice solution would be to ensure that scaling, translating and rotation properly update the corresponding arc attributes (`scale`, `start_angle`, `arc_center`), and then `inner_arc` and `outer_arc` could be implemented as computed properties, like

```py
@property
def inner_arc(self):
    return Arc(radius=self.radius, arc_center=self.arc_center, ...)
```

However, just from scrolling over `mobject.py` I am not sure whether simply overriding `scale`, `shift`, and `rotate` is enough, especially considering that something like `stretch` is a thing. For these, something easy like

```py
def scale(self, scale_factor, **kwargs):
    self.radius *= scale_factor
    super().scale(scale_factor, **kwargs)
```

should be sufficient.

(Problems introduced by `stretch`, that is, a transformation that changes the nature of mobjects, are probably a different story, though.)

If you'd like to try something like that you would be very welcome -- otherwise we can move cleaning this up to a follow-up PR.

**Comment by AkashKarnatak** [comment]
To implement `inner_arc` and `outer_arc` functions, we just need to find `arc_center`, `inner_radius` and `outer_radius`. 

`arc_center` can be found using the `get_arc_center` method.
`inner_radius` is the distance between `arc_center` and `self.points[0]`(first anchor point of of inner arc) and `outer_radius` is the distance between `arc_center` and `self.points[(self.num_components-1)*4+3]` (first anchor point of outer arc)
Also, `scale` of annular sector is equal to `outer_radius_at_any_instant/initial_outer_radius` and `rotation` can be found by calculating the difference between the angle subtended by line joining `arc_center` and first anchor of `outer_arc` and `self.start_angle + self.angle`

I guess this way we can keep track of basic transformations.

File: tests/test_geometry.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import pytest
```
Comment (huguesdevimeux) [inline tests/test_geometry.py:1]: Unfortunatly there is already a test_geometry file, which is for graphical unit tests of geometrey module. Could you please rename this file into something like "test_unit_geometry.py'" ?

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]

File: tests/test_geometry.py (hunk 1 above)
Comment (AkashKarnatak) [reply tests/test_geometry.py]: Done

**Comment by behackl** [comment]
> To implement `inner_arc` and `outer_arc` functions, we just need to find `arc_center`, `inner_radius` and `outer_radius`.
> 
> `arc_center` can be found using the `get_arc_center` method.
> `inner_radius` is the distance between `arc_center` and `self.points[0]`(first anchor point of of inner arc)
> `outer_radius` is the distance between `arc_center` and `self.points[(self.num_components-1)*4+3]` (first anchor point of outer arc)
> 
> Also, `scale` of annular sector is equal to `outer_radius_at_any_instant/initial_outer_radius` and `rotation` can be found by calculating the difference between the angle subtended by line joining `arc_center` and first anchor of `outer_arc` and `self.start_angle + self.angle`
> 
> I guess this way we can keep track of basic transformations.

Yes, this is another approach that seems to require less refactoring. So, either keep track of the arc characteristics by overriding `scale`, `shift`, `rotate`, or by computing everything from the points. I am somewhat indifferent to which approach I like more, maybe I have a slight tendency towards the "overriding" version (but just because I feel that this is how I would do it, if I would write it from scratch and because I have a slight dislike for point computations).

Anyhow, I start to get the impression that it might be better to move further discussions regarding refactoring `Arc` and its derived classes to a separate issue (otherwise #480 might remain a problem for quite some more time...). I suggest that we do that.

File: tests/test_unit_geometry.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import pytest
+import numpy as np
+from manim import Sector, ORIGIN
+
+
+def test_get_arc_center():
+    assert np.all(Sector().get_arc_center() == ORIGIN)
```
Comment (behackl) [inline tests/test_unit_geometry.py:7]: This test also passes without your patch to the function: if the intersection in `get_arc_center` fails, then it always returns `ORIGIN`. Please add a test for a sector/arc whose center is not `ORIGIN`.

**Review by behackl — CHANGES_REQUESTED** [review]

**Comment by AkashKarnatak** [comment]
> Yes, this is another approach that seems to require less refactoring. So, either keep track of the arc characteristics by overriding `scale`, `shift`, `rotate`, or by computing everything from the points. I am somewhat indifferent to which approach I like more, maybe I have a slight tendency towards the "overriding" version (but just because I feel that this is how I would do it, if I would write it from scratch and because I have a slight dislike for point computations).
> 
> Anyhow, I start to get the impression that it might be better to move further discussions regarding refactoring `Arc` and its derived classes to a separate issue (otherwise #480 might remain a problem for quite some more time...). I suggest that we do that.

Okay, let's move our discussion over to a separate issue. Maybe, I'll open one today.

**Review by behackl — APPROVED** [review]
LGTM.

**Review by leotrs — APPROVED** [review]

**Pull request merged**